python/lsst/ts/scriptqueue/ui/request_model.py

```diff
@@ -0,0 +1,547 @@
+            print("Enabling the queue.")
+        else:
+            print(f"Queue in {salobj.State(state.summaryState)} state, cannot enable.")
+
+        self.run(self.queue.cmd_start.start(timeout=self.cmd_timeout))
+        self.run(self.queue.cmd_enable.start(timeout=self.cmd_timeout))
+        print(f"Queue enabled. [{self.queue.evt_summaryState.get().summaryState}]")
+
+        return False
+
+    def list(self):
+        """Return available scripts.
+
+        Returns
+        -------
+        available_scripts : dict
+            A dictionary with "external" and "standard" list of scripts.
+
+        """
+
+        script_list = self.queue.evt_availableScripts.get()
+        if script_list is None:
+            self.log.debug("No previous information about script list. Requesting.")
+            available_scripts_coro = self.queue.evt_availableScripts.next(flush=False,
+                                                                          timeout=self.cmd_timeout)
+
+            self.run(self.queue.cmd_showAvailableScripts.start(timeout=self.cmd_timeout))
+
+            script_list = self.run(available_scripts_coro)
+
+        return {'external': script_list.external.split(':'),
+                'standard': script_list.standard.split(':')}
+
+    def pause_queue(self):
+        """Pause the queue."""
+        self.run(self.queue.cmd_pause.start())
+
+    def resume_queue(self):
+        """Resume queue operation"""
+        self.run(self.queue.cmd_resume.start())
+
+    def quit_queue(self):
+        """Quit queue by sending it to exit control. Assume the queue is in enable state
+        but if it is not, will skip the rejected commands for disable and standby.
+        """
+
+        try:
+            self.run(self.queue.cmd_disable.start(timeout=self.cmd_timeout))
+        except Exception as e:
+            pass
+
+        try:
+            self.run(self.queue.cmd_standby.start(timeout=self.cmd_timeout))
+        except Exception as e:
+            pass
+
+        try:
+            self.run(self.queue.cmd_exitControl.start(timeout=self.cmd_timeout))
+        except Exception as e:
+            pass
+
+        return True
+
+    def stop_scripts(self, sal_indices, terminate=False):
+        """Stop list of scripts.
+
+        Parameters
+        ----------
+        sal_indices : list(int)
+            A list of script indices to stop.
+
+        terminate : bool
+            Should the script be terminated?
+
+        """
+        self.queue.cmd_stopScripts.set(salIndices=sal_indices,
+                                       length=len(sal_indices),
+                                       terminate=terminate)
+        self.run(self.queue.cmd_stopScripts.start(timeout=self.cmd_timeout))
+
+    def get_queue_state(self):
+        """A method to get the state of the queue and the scripts running on the queue.
+
+        Returns
+        -------
+        state : dict
+
+        """
+        self.update_queue_state()
```

Comment:
I strongly suggest that you make update_queue_state a callback for self.queue.evt_queue and have it update self.state accordingly (though it will have to take an argument -- the new queue state). Then `get_queue_state` can just return `self.state`.

Also consider renaming "self.state" to "self._queue_state" and making this a property `queue_state` that simply returns that attribute.

Reply:
Yes, I thought about that.. I really wanted to have this update as a callback but unfortunately the `cmd` library does not play well with asyncio. I really tried...